### PR TITLE
planner: fix the issue of reusing wrong point-plan for "select ... for update"

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #54652

Problem Summary: planner: fix the issue of reusing wrong point-plan for "select ... for update"

### What changed and how does it work?

Encode more txn state into the plan cache key, and check whether the key has changed before reusing point-get plans.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by cfzjywxk — APPROVED** [review]
LGTM, some refactor and improvements mentioned by @MyonKeminta could be done in later PRs.

File: pkg/planner/core/plan_cache_utils.go [hunk 1]
```diff
@@ -386,9 +383,23 @@ func NewPlanCacheKey(sctx sessionctx.Context, stmt *PlanCacheStmt) (key, binding
 			hash = codec.EncodeInt(hash, id)
 		}
 	}
+
+	// txn status
+	hash = append(hash, '|')
+	hash = append(hash, bool2Byte(vars.InTxn()))
+	hash = append(hash, bool2Byte(vars.IsAutocommit()))
+	hash = append(hash, bool2Byte(config.GetGlobalConfig().PessimisticTxn.PessimisticAutoCommit.Load()))
```
Comment (AilinKid) [inline pkg/planner/core/plan_cache_utils.go:391]: we only cache the txn STATE info here, not the real txn start timestamp, is this enough?

Comment (AilinKid) [reply pkg/planner/core/plan_cache_utils.go:391]: oh, my bad, cachable plan mustn't care about txn start ts

**Review by AilinKid — APPROVED** [review]

File: pkg/ddl/tests/metadatalock/mdl_test.go [hunk 2]
```diff
@@ -935,7 +935,7 @@ func TestMDLPreparePlanCacheExecute(t *testing.T) {
 	tk.MustQuery("select * from t2")
 	tk.MustExec(`set @a = 2, @b=4;`)
 	tk.MustExec(`execute stmt_test_1 using @a, @b;`)
-	tk.MustQuery("select @@last_plan_from_cache;").Check(testkit.Rows("1"))
+	tk.MustQuery("select @@last_plan_from_cache;").Check(testkit.Rows("0"))
```
Comment (MyonKeminta) [inline pkg/ddl/tests/metadatalock/mdl_test.go:938]: Should we also have a separated test case for the case about preparing `select ... for update` statement outside transaction? This test targets on testing MDL, and I'm afraid that this test might being modified / removed without being known by us and making the case lose coverage...

**Review by MyonKeminta — COMMENTED** [review]
Rest LGTM

File: pkg/ddl/tests/metadatalock/mdl_test.go (hunk 2 above)
Comment (qw4990) [reply pkg/ddl/tests/metadatalock/mdl_test.go:940]: I enhancemented this test, PTAL

**Review by MyonKeminta — APPROVED** [review]

File: pkg/ddl/tests/metadatalock/mdl_test.go [hunk 3]
```diff
@@ -934,7 +934,11 @@ func TestMDLPreparePlanCacheExecute(t *testing.T) {
 
 	tk.MustQuery("select * from t2")
 	tk.MustExec(`set @a = 2, @b=4;`)
-	tk.MustExec(`execute stmt_test_1 using @a, @b;`)
+	tk.MustExec(`execute stmt_test_1 using @a, @b;`) // can't reuse the prior plan created outside this txn.
+	tk.MustQuery("select @@last_plan_from_cache;").Check(testkit.Rows("0"))
+	tk.MustExec(`execute stmt_test_1 using @a, @b;`) // can't reuse the prior plan since this table becomes dirty.
+	tk.MustQuery("select @@last_plan_from_cache;").Check(testkit.Rows("0"))
+	tk.MustExec(`execute stmt_test_1 using @a, @b;`) // can't reuse the prior plan now.
```
Comment (MyonKeminta) [inline pkg/ddl/tests/metadatalock/mdl_test.go:941]:
```suggestion
	tk.MustExec(`execute stmt_test_1 using @a, @b;`) // can reuse the prior plan now.
```
Did you mean this?

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**